File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int>postfix(n);
        vector<int>prefix(n);
        prefix[0] = nums[0];
        for(int i=1; i<n; i++) {
            prefix[i] = nums[i] * prefix[i-1];
        }
        postfix[n-1] = nums[n-1];
        for(int i=n-2; i>=0; i--) {
            postfix[i] = postfix[i+1] * nums[i];
        }
        nums[0] = postfix[1];
        nums[n-1] = prefix[n-2];
        for(int i=1; i<n-1; i++) {
            nums[i] = postfix[i+1] * prefix[i-1];
        }
        return nums;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        p = 1
        z1 = False
        a = [0] * len(nums)

        
        for i in range(len(nums)):
            if nums[i] == 0 and z1:
                return a
            if nums[i] == 0 and z1 == False:
                z1 = True
                zi = i
                continue
        
            
            p *= nums[i]
        
        
        if z1:
            a[zi] = p
            return a

        
        
        for i in range(len(nums)):
            if z1 and not nums[i]:
                a[i] = p
                return a
            
            tmp = nums[i]
            
            a[i] = int(p/nums[i])
                
        return a